---
title: "R_project_McMurchie"
output: html_document
date: '2022-03-10'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
# McMurchie R Assignment

Read in necessary libraries first
```{r}
library(tidyverse)
library(dplyr)
library(janitor) # helps in data cleanup
library(data.table) # specifically for data table work
library(stringr)
library(ggpubr) # used for gg.arrange function (arranges plots)
library(RColorBrewer) # for plot colors

```


## Data Inspection

### Attributes of `fang_et_al_genotypes.txt`

```{r}
geno <- read.delim("https://raw.githubusercontent.com/EEOB-BioData/BCB546-Spring2022/main/assignments/UNIX_Assignment/fang_et_al_genotypes.txt", header = TRUE, sep = "\t")
# read in fang_et_al_genotypes.txt 

view(geno) # allows us to visually inspect file
# We see that there is a header line with Sample_ID, JG_OTU, and Group
# These three columns are followed by a number of columns labeled with SNP ID
# This header line is followed by the data

ncol(geno) # count number of columns
# 986

nrow(geno) # count number of rows
# 2782 (excludes header)

dim(geno) # count dimensions (rows and columns)
# 2782 rows (this excludes header - with header, 2783 rows), 986 columns
# Multiplying 2782 by 986 gives the number of "words" excluding header: 2,743,052
# Multiplying 2783 by 986 gives the number of "words" including header: 2,744,038

sum(!is.na(geno)) # this will give word count excluding header
# 2743052

length(geno) # number of columns
# 986

object.size(geno) # size in bytes
# 22681376 bytes

str(geno) # structure of geno
# dataframe, 2782 obs (rows excluding header). of 986 variables (columns)
# Although list output is truncated, first few variables read as characters

```
We see that there is a header line with Sample_ID, JG_OTU, and Group, with three columns are followed by a number of columns labeled with SNP ID (still part of the header). There are 986 columns and 2782 rows, exluding the header. Excluding header, there are 2,743,052 "words" (essentially, entries), while including header, there are 2,744,038. The size of the object is 22681376 bytes. Once imported, `fang_et_al_genotypes.txt` is structured as a dataframe, with the first few variables read as characters.

### Attributes of `snp_position.txt`

```{r}
snp_pos <- read.delim("https://raw.githubusercontent.com/EEOB-BioData/BCB546-Spring2022/main/assignments/UNIX_Assignment/snp_position.txt", header = TRUE, sep = "\t")
# read in snp_position.txt

view(snp_pos) # allows us to visually inspect file
# We see that there is a header line with SNP_ID,
# Chromosome, Position, alt_pos, mult_positions, amplicon,
# cdv_map_feature.name, gene,and candidate.random, cvd_marker_id,
# Genaissance_daa_id, Sequenom_daa_id, count_amplicons, count_cmf,and count_gene
# The header is followed by the data

ncol(snp_pos) # count number of columns
# 15 

nrow(snp_pos) # count number of rows
# 983 (excludes header)

dim(snp_pos) # count dimensions (rows and columns)
# 983 rows (this excludes header), 15 columns
# Multiplying 983 by 15 gives the number of "words" excluding header: 14,745
# Multiplying 984 by 15 gives the number of "words" including header: 14,760

sum(!is.na(snp_pos)) # this will give word count excluding header
# 14745

length(snp_pos) # number of columns
# 15

object.size(snp_pos) # size in bytes
# 327392 bytes

str(snp_pos) # structure of snp_pos
# dataframe, 983 obs (rows excluding header). of 15 variables (columns)
# SNP_ID, Chromosome, Position, alt_pos, mult_positions, amplicon,
# cdv_map_feature.name, gene,and candidate.random are characters
# cvd_marker_id, Genaissance_daa_id, Sequenom_daa_id, count_amplicons, count_cmf,
# and count_gene are integers

```
We see that there is a header line with SNP_ID, Chromosome, Position, alt_pos, mult_positions, amplicon, cdv_map_feature.name, gene,and candidate.random, cvd_marker_id, Genaissance_daa_id, Sequenom_daa_id, count_amplicons, count_cmf,and count_gene. The header is followed by the data. There are 15 columns and 983 rows, exluding the header. Excluding header, there are 14,745 "words" (essentially, entries), while including header, there are 14,760. The size of the object is 14745 bytes. Once imported, `fang_et_al_genotypes.txt` is structured as a dataframe, with the first few variables read as characters.

## Data Processing

### Initial Data Processing Steps

```{r}
snp_cut <- snp_pos %>%
  select(SNP_ID, Chromosome, Position) # select only SNP_ID, Chromosome, and Position

```
Here, we selected only the SNP_ID, Chromosome, and Position columns from `snp_pos` and assigned them to the new `snp_cut` dataframe.

### Maize Files Processing

```{r}
geno_maize <- geno %>%
  filter(Group %in% c("ZMMIL", "ZMMLR", "ZMMMR")) %>%
  # filter by maize groups only, including sample ID
  t() %>%
  # transpose - note it's added automatic column names (V1, V2, etc. in new header)
  as.data.frame() %>%
  # make sure it's a dataframe
  setDT(keep.rownames = TRUE) %>%
  #gives column 1 a placeholder name as well
  row_to_names(row_number = 1) %>%
  # sets first row as column names
  rename(SNP_ID = Sample_ID)
# rename first column as SNP_ID so we can join on like column names

join_maize <- snp_cut %>%
  inner_join(geno_maize, by = "SNP_ID")
# joins both files - inner_join makes sure we take only SNP_IDs in both files

sort_maize <- join_maize %>%
  arrange(as.numeric(Chromosome), as.numeric(Position))
# sorts by chromosome number and position, in ascending order

for (i in 1:10) {
  temp_df <- sort_maize %>%
    filter(Chromosome == i) 
write.table(temp_df, file = paste("./final_files/final_maize_files/maize_chr",i,".tsv", sep = ""), row.names = FALSE, col.names = TRUE)
}
# this step splits sort_maize into ten files (based on a temporary dataframe) one for each chromosome number

## Below, steps for making files with descending position and dash

sort_desc_maize <- join_maize %>%
  arrange(as.numeric(Chromosome), desc(as.numeric(Position))) 
# sorts maize file in ascending order by chromosome and descending order by position

maize_dash <- data.frame(lapply(sort_desc_maize, gsub, pattern = "[?]", replacement = "-"))
# replaces ? with -

for (i in 1:10) {
  temp_df <- maize_dash %>%
    filter(Chromosome == i) 
write.table(temp_df, file = paste("./final_files/final_maize_files/maize_dash_chr",i,".tsv", sep = ""), row.names = FALSE, col.names = TRUE)
}
# this step splits maize_dash into ten files (based on a temporary dataframe) one for each chromosome number

```

First, I created the `geno_maize` data frame from `geno` by filtering only maize groups. I transposed so that I could join this object to `snp_cut` later, and set it as a dataframe, giving column 1 a placeholder name. I then set the first row as the column names and renamed SNP_ID as Sample_Id to facilitate joining.

My next step was to join the `geno_maize` and `snp_cut` dataframes to make `join_maize`. 

I then sorted `join_maize` by Chromosome number and Position, in ascending order, in `sort_maize`. 

At this point, I split `sort_maize` into 10 files, one for each chromosome, which contained the SNP positions associated with each chromosome.

To make my files sorted with position in descending order, I returned to `join_maize` and left Chromosome sorted in ascending order while sorting Position in descending order in `sort_desc_maize`. Finally, I replaced the "?" indicating missing data with a "-" in `maize_dash` and split this file into 10 files, one for each chromosome.  


### Teosinte Files Processing

```{r}
geno_teosinte <- geno %>%
  filter(Group %in% c("ZMPBA", "ZMPIL", "ZMPJA")) %>% 
  # filter by maize groups only, including sample ID
  t() %>%
  # transpose - note it's added automatic column names (V1, V2, etc. in new header)
  as.data.frame() %>%
  # make sure it's a dataframe
  setDT(keep.rownames = TRUE) %>%
  #gives column 1 a placeholder name as well
  row_to_names(row_number = 1) %>%
  # sets first row as column names
  rename(SNP_ID = Sample_ID)
# rename first column as SNP_ID so we can join on like column names

join_teosinte <- snp_cut %>%
  inner_join(geno_teosinte, by = "SNP_ID")
# joins both files - inner_join makes sure we take only SNP_IDs in both files

sort_teosinte <- join_teosinte %>%
  arrange(as.numeric(Chromosome), as.numeric(Position))
# sorts by chromosome number and position, in ascending order

for (i in 1:10) {
  temp_df <- sort_teosinte %>%
    filter(Chromosome == i) 
write.table(temp_df, file = paste("./final_files/final_teosinte_files/teosinte_chr",i,".tsv", sep = ""), row.names = FALSE, col.names = TRUE)
}
# this step splits sort_teosinte into ten files (based on a temporary dataframe) one for each chromosome number

sort_desc_teosinte <- join_teosinte %>%
  arrange(as.numeric(Chromosome), desc(as.numeric(Position))) 
# sorts teosinte file in ascending order by chromosome and descending order by position

teosinte_dash <- data.frame(lapply(sort_desc_teosinte, gsub, pattern = "[?]", replacement = "-"))
# replaces ? with -

for (i in 1:10) {
  temp_df <- teosinte_dash %>%
    filter(Chromosome == i) 
write.table(temp_df, file = paste("./final_files/final_teosinte_files/teosinte_dash_chr",i,".tsv", sep = ""), row.names = FALSE, col.names = TRUE)
}
# this step splits teosinte_dash into ten files (based on a temporary dataframe) one for each chromosome number

```

For the teosinte data, I followed a similar process as I did for the maize data, first creating the `geno_teosinte` data frame from `geno` by filtering only teosinte groups. I transposed so that I could join this object to `snp_cut` later, and set it as a dataframe, giving column 1 a placeholder name. I then set the first row as the column names and renamed SNP_ID as Sample_Id to facilitate joining.

I then joined `geno_teosinte` and `snp_cut` to make `join_teosinte`. 

Next, I sorted`join_teosinte` by Chromosome number and Position, in ascending order, in `sort_teosinte`. 

To make my 10 teosinte files with chomrosome and position in ascending order, I split `sort_teosinte` into 10 files, one for each chromosome.

To make my files sorted with position in descending order, I went back to `join_teosinte` and left Chromosome sorted in ascending order while sorting Position in descending order in `sort_desc_teosinte`. Finally, I replaced the "?" indicating missing data with a "-" in `teosinte_dash` and split this file into 10 files, one for each chromosome. 

## Data Visualization

### SNPs per chromosome

```{r}
sort_maize_bin <- sort_maize %>% 
  filter(Position != "unknown") %>%
  filter(Position != "multiple") %>% # keep only chromosomes of known position
  mutate(position_binned = cut(as.numeric(Position), 10)) %>% # make 10 bins per chromosome
  filter(Chromosome != "multiple") %>% 
  filter(Chromosome != "unknown") %>% # keep only SNPs where chromosome is known
  mutate(Chromosome = fct_relevel(Chromosome, c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10")))
# final step sets levels so graph reads in order of chromosome number

maize <- ggplot(data = sort_maize_bin) + 
  geom_bar(mapping = aes(x = position_binned, fill = Chromosome), color = "black") + # colored by chromosome, with black outline
  xlab("Position(basepairs)") +
  ylab("SNP Distribution") +
  theme(axis.text.x = element_text(angle = 90)) + # changes angle so position is readable
  scale_fill_brewer(palette = "Spectral") +
  facet_grid(~as.numeric(Chromosome)) # facet by chromosome


sort_teosinte_bin <- sort_teosinte %>% 
  filter(Position != "unknown") %>%
  filter(Position != "multiple") %>% # keep only chromosomes of known position
  mutate(position_binned = cut(as.numeric(Position), 10)) %>% # make 10 bins per chromosome
  filter(Chromosome != "multiple") %>%
  filter(Chromosome != "unknown") %>% # keep only SNPs where chromosome is known
  mutate(Chromosome = fct_relevel(Chromosome, c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10")))
# final step sets levels so graph reads in order of chromosome number

teosinte <- ggplot(data = sort_teosinte_bin) + 
  geom_bar(mapping = aes(x = position_binned, fill = Chromosome), color = "black") + # colored by chromosome, with black outline
  xlab("Position(basepairs)") +
  ylab("SNP Distribution") +
  theme(axis.text.x = element_text(angle = 90)) + # changes angle so position is readable
  scale_fill_brewer(palette = "Spectral") + # change plot colors (from RColorBrewer)
  facet_grid(~as.numeric(Chromosome)) # facet by chromosome

SNP_figure <- ggarrange(maize, teosinte, # from ggpubr, arranges graphs
                    labels = c("maize", "teosinte"),
                    ncol = 1, nrow = 2) # graphs arranged with maize on top, teosinte below
SNP_figure

ggsave("./plots/SNP_figure.pdf", plot = SNP_figure, width = 12, height = 15, units = "in", dpi = 300) # print to pdf

ggsave("./plots/SNP_figure.png", plot = SNP_figure, width = 12, height = 15, units = "in", dpi = 300) # print to png


```

Here, I made a graph showing the positions of each SNP on each chromosome in both maize and teosinte. 

To do this, I first filtered out the SNPs that were in unknown and multiple positions and made 10 bins per chromosome, filtering out SNPs where chromosome isn't known. I then changed the chromosome numbers to factors and changed their order so that they would read 1-10, instead of 1, 10, 2, 3, etc. When I did this for maize, I put this into `sort_maize_bin`. For teosinte, I put the results into `sort_teosinte_bin`. 

Then, for both maize and teosinte, I made a plot of position binned, with position colored and faceted by chromosome. 

I arranged these two plots into a single figure, `SNP_figure`, using `ggarrange` from the ggpubr package, and saved them as a pdf and png. Note: the pdf and png versions of this figure are easier to read than the version in the html due to increased horizontal dimensions.

From this figure, we can see that there are similar numbers of SNP positions on each chromosome for maize and teosinte, but that on each chromosome, there are fewer SNP positions in the middle (the centromere). Some chromosomes have more SNP positions than others. For example, chromosome 1 has many more SNP positions than Chromosome 10. 

```{r}
maize_unknown <- sort_maize %>%
  filter(Chromosome == "unknown")
# makes datafame for maize SNPs on unknown chromosome

maize_multiple <- sort_maize %>%
  filter(Chromosome == "multiple")
# makes datafame for maize SNPs on unknown chromosome

Chromosome_split_maize <- split(sort_maize, sort_maize$Chromosome)
# splits sort_maize by chromosome number

maize_chr_data <- c("maize_chr1", "maize_chr10", "maize_chr2", "maize_chr3", "maize_chr4", "maize_chr5", "maize_chr6", "maize_chr7", "maize_chr8", "maize_chr9")
# makes a list of names for dataframes for each chromosome of maize

for (i in 1:length(Chromosome_split_maize)) {assign(maize_chr_data[i], Chromosome_split_maize[[i]])}
# a loop to split the data based upon the created group of lists (Chromosome_split_maize) and assigning the corresponding name as the object.

mchr1 <- nrow(maize_chr1)
mchr2 <- nrow(maize_chr2)
mchr3 <- nrow(maize_chr3)
mchr4 <- nrow(maize_chr4)
mchr5 <- nrow(maize_chr5)
mchr6 <- nrow(maize_chr6)
mchr7 <- nrow(maize_chr7)
mchr8 <- nrow(maize_chr8)
mchr9 <- nrow(maize_chr9)
mchr10 <- nrow(maize_chr10)
mchrmult <- nrow(maize_multiple)
mchrunkn <- nrow(maize_unknown)
# counts number of SNP positions on each chromosome and on multiple and unknown chromosomes

teosinte_unknown <- sort_teosinte %>%
  filter(Chromosome == "unknown")

teosinte_multiple <- sort_teosinte %>%
  filter(Chromosome == "multiple")

Chromosome_split_teosinte <- split(sort_teosinte, sort_teosinte$Chromosome)
# splits sort_teosinte by chromosome number

teosinte_chr_data <- c("teosinte_chr1", "teosinte_chr10", "teosinte_chr2", "teosinte_chr3", "teosinte_chr4", "teosinte_chr5", "teosinte_chr6", "teosinte_chr7", "teosinte_chr8", "teosinte_chr9")
# makes a list of names for dataframes for each chromosome of teosinte

for (i in 1:length(Chromosome_split_teosinte)) {assign(teosinte_chr_data[i], Chromosome_split_teosinte[[i]])}
# a loop to split the data based upon the created group of lists (Chromosome_split_teosinte) and assigning the corresponding name as the object.

tchr1 <- nrow(teosinte_chr1)
tchr2 <- nrow(teosinte_chr2)
tchr3 <- nrow(teosinte_chr3)
tchr4 <- nrow(teosinte_chr4)
tchr5 <- nrow(teosinte_chr5)
tchr6 <- nrow(teosinte_chr6)
tchr7 <- nrow(teosinte_chr7)
tchr8 <- nrow(teosinte_chr8)
tchr9 <- nrow(teosinte_chr9)
tchr10 <- nrow(teosinte_chr10)
tchrmult <- nrow(teosinte_multiple)
tchrunkn <- nrow(teosinte_unknown)
# counts number of SNP positions on each chromosome and on multiple and unknown chromosomes

SNP_pos_df <- data.frame(chromosome = c("chr 1", "chr 2", "chr 3", "chr 4", "chr 5", "chr 6", "chr 7", "chr 8", "chr 9", "chr 10", "multiple", "unknown", "chr 1", "chr 2", "chr 3", "chr 4", "chr 5", "chr 6", "chr 7", "chr 8", "chr 9", "chr 10", "multiple", "unknown"),
                    species = c("maize", "maize", "maize", "maize", "maize", "maize", "maize", "maize", "maize", "maize", "maize", "maize", "teosinte", "teosinte", "teosinte", "teosinte", "teosinte", "teosinte", "teosinte", "teosinte", "teosinte", "teosinte", "teosinte", "teosinte"),
                    SNP_positions = c(mchr1, mchr2, mchr3, mchr4, mchr5, mchr6, mchr7, mchr8, mchr9, mchr10, mchrmult, mchrunkn, tchr1, tchr2, tchr3, tchr4, tchr5, tchr6, tchr7, tchr8, tchr9, tchr10, tchrmult, tchrunkn)) %>%
  mutate(chromosome = fct_relevel(chromosome, c("chr 1", "chr 2", "chr 3", "chr 4", "chr 5", "chr 6", "chr 7", "chr 8", "chr 9", "chr 10", "multiple", "unknown")))
# making dataframe with each chromosome, speices, and number of SNP positions. Final step (with factor relevel) sets chromosome as factor and changes order so that they read 1, 2, 3, etc. rather than 1, 10, 2, etc. 

SNP_pos_bar <- ggplot(data = SNP_pos_df, aes(x = species, y = SNP_positions, fill = species)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  theme(axis.text.x = element_text(angle = -90)) + # makes species labels vertical so they're legible 
  ylim(0,170) + # increases y-axis limit so text labels are visible
   geom_text(aes(label = SNP_positions), angle = -90, hjust = 1.2) + # adds numeric text labels above bars
  xlab("Species") +
  ylab("Number of SNP positions") +
  scale_fill_brewer(palette = "Set1") + # Colors from RColorBrewer
  facet_grid(~chromosome) # facet by chromosome
SNP_pos_bar

ggsave("./plots/SNP_pos_bar.pdf", plot = SNP_pos_bar, width = 10, height = 5, units = "in", dpi = 300) # print to pdf

ggsave("./plots/SNP_pos_bar.png", plot = SNP_pos_bar, width = 10, height = 5, units = "in", dpi = 300) # print to png

```

Using these steps, I made a graph showing the total number of SNP positions for both teosinte and maize on each chromosome and on unknown or multiple chromosomes.

To do this, I first made objects containing the SNP positions on unknown and multiple chromosomes for both teosinte and maize. Next, I split the sorted maize and teosinte dataframes based on chromosome number, made a list containing names for each chromosome for each species, to use with the dataframe I'd made previously. To combine these, I ran a loop that assigned the data for each chromosome from each species to the corresponding name from the list.

I then counted the number of rows (corresponding to number of SNP positions) for chromosome (and unknown and multiple chromosomes) for each species and put this number into an object. I combined these objects, along with the species names and chromosome numbers, to make a new dataframe `SNP_pos_df`, also making the chromosome numbers factors and changing their order so that it would display as 1, 2, 3, etc. rather than 1, 10, 2. 

I then made my plot, setting x to species and y to the number of SNP positions, saving the graph as both a png and pdf. 

This graph shows that the number of SNP positions is identical for each chromosome (and unknown and multiple chromosomes) for both maize and teosinte.

### Missing data and amount of heterozygosity

```{r}
geno5 <- geno %>%
  mutate(homozygous = apply(geno, 1, function(x) length(which(x == ("A/A") | x == ("T/T") | x == ("C/C") | x == ("G/G"))))) %>%
  mutate(heterozygous = apply(geno, 1, function(x) length(which(x == ("A/C") | x == ("A/T") | x == ("A/G") | x == ("C/A") | x == ("C/T") | x == ("C/G") | x == ("G/A") | x == ("G/C") | x == ("G/T") | x == ("T/A") | x == ("T/C") | x == ("T/G"))))) %>%
  mutate(missing = apply(geno, 1, function(x) length(which(x == ("?/?"))))) %>%
  mutate(total =  apply(geno, 1, function(x) length(which(x == ("?/?") | x == ("A/A") | x == ("T/T") | x == ("C/C") | x == ("G/G") | x == ("A/C") | x == ("A/T") | x == ("A/G") | x == ("C/A") | x == ("C/T") | x == ("C/G") | x == ("G/A") | x == ("G/C") | x == ("G/T") | x == ("T/A") | x == ("T/C") | x == ("T/G")))))
# makes dataframe with new columns for homozygous SNP positions, heterozygous SNP positions, missing data, and total

geno_totals <- geno5 %>%
  filter(Group %in% c("ZMMIL", "ZMMLR", "ZMMMR", "ZMPBA", "ZMPIL", "ZMPJA")) %>%
  select(Sample_ID, Group, homozygous, heterozygous, missing) %>%
  mutate(species = fct_recode(Group, 
                             "maize" = "ZMMIL", 
                             "maize" = "ZMMLR",
                             "maize" = "ZMMMR",
                             "teosinte" = "ZMPBA",
                             "teosinte" = "ZMPIL",
                             "teosinte" = "ZMPJA")) 
# filters geno5 so that only maize and teosinte groups are selected, as well as only the Sample_ID, Group, and zygosity and missing columns. Also adds a species column.

big_zygo_pivot <- geno_totals %>% 
  pivot_longer(., cols = c(homozygous, heterozygous, missing), names_to = "zygosity", values_to = "count")
# pivots the zygosity and missing data columns into two columns: zygosity and count

big_zygo_species_plot <- ggplot(big_zygo_pivot, aes(fill = zygosity, y = count, x = species)) + 
  xlab("Species") +
  ylab("Proportion") +
  geom_bar(position = "fill", stat = "identity") +
  scale_fill_brewer(palette = "Set1") # colors from RColorBrewer
big_zygo_species_plot


ggsave("./plots/zygo_species_plot.pdf", plot = big_zygo_species_plot, width = 5, height = 5, units = "in", dpi = 300) # print to pdf

ggsave("./plots/zygo_species_plot.png", plot = big_zygo_species_plot, width = 5, height = 5, units = "in", dpi = 300) # print to png

```

With these steps, I made a graph displaying the proportion of homozygous, heterozygous, and missing SNP positions for each species (maize and teosinte).

I started by adding in columns that counted up the total number of homozygous sites, heterozygous sites, and sites with missing data. Then I filtered to include only maize and teosinte groups, selected only the Group, homozygous, heterozygous, and missing count columns, and added a column for species, putting this in `geno_totals`. I pivoted the homozygous, heterozygous, and missing columns to make a longer zygosity column, with values in a count column in `big_zygo_pivot`. 

From this, I made a graph showing the proportion of homozygous, heterozygous, and missing SNP positions for both maize and teosinte, saving this graph as a pdf and png. We can see that maize has very slightly more missing positions, while teosinte has more heterozygous SNP positions.

```{r}
big_zygo_group_plot <- ggplot(big_zygo_pivot, aes(fill = zygosity, y = count, x = Group)) + 
  xlab("Group") +
  ylab("Proportion") +
  geom_bar(position = "fill", stat = "identity") +
  theme(axis.text.x = element_text(angle = 90)) + # add rotated text for group names
  scale_fill_brewer(palette = "Set1") + # colors from RColorBrewer
  facet_grid(~species) # facet by species
big_zygo_group_plot

ggsave("./plots/zygo_group_plot.pdf", plot = big_zygo_group_plot, width = 5, height = 5, units = "in", dpi = 300) # print to pdf

ggsave("./plots/zygo_group_plot.png", plot = big_zygo_group_plot, width = 5, height = 5, units = "in", dpi = 300) # print to png

```

The above graph shows the proportion of homozygous, heterozygous, and missing SNP positions for each group, faceted by species, with the graph saved as a pdf and png. 

The groups ZMPBA, ZMPIL, and ZMPJA are empty for maize because these were teosinte groups. The groups ZMMIL, ZMMLR, and ZMMMR were empty for teosinte because these were maize groups. 

From this graph, we can see that missing data is slightly higher in the maize groups ZMMIL and ZMMMR than in other groups. Heterzygosity is extremely low in the maize group ZMMIL, and fairly low in maize group ZMMMR and teosinte group ZMPIL compared to in other groups.

```{r}
big_zygo_sample_plot <- ggplot(big_zygo_pivot, aes(fill = zygosity, y = count, x = Sample_ID)) +
  xlab("Sample") +
  ylab("Proportion") +
  geom_bar(position = "fill", stat = "identity") +
  scale_fill_brewer(palette = "Set1") # colors from RColorBrewer
big_zygo_sample_plot 

ggsave("./plots/zygo_sample_plot.pdf", plot = big_zygo_sample_plot, width = 15, height = 5, units = "in", dpi = 300) # print to pdf

ggsave("./plots/zygo_sample_plot.png", plot = big_zygo_sample_plot, width = 15, height = 5, units = "in", dpi = 300) # print to png
```

This shows the proportion of homozygous, heterozygous, and missing SNP positions for each sample, with the graph saved as a pdf and png. Due to the very high number of samples, Sample ID is not legible. 

From this graph, we can see that the proportion of missing data is similar for most samples. Heterozygosity is very high for a few samples (maybe about 1/5 of the samples), but similar in proportion among most other samples.

### Graph of our choice
```{r}
geno8 <- geno %>%
  mutate(GC = apply(geno, 1, function(x) length(which(x == ("G/C") | x == ("C/G") | x == ("G/G") | x == ("C/C"))))) %>%
  mutate(not_GC = apply(geno, 1, function(x) length(which(x == ("?/?") | x == ("A/A") | x == ("T/T") | x == ("A/C") | x == ("A/T") | x == ("A/G") | x == ("C/A") | x == ("C/T") | x == ("G/A") | x == ("G/T") | x == ("T/A") | x == ("T/C") | x == ("T/G")))))
# adds two new columns, one (GC) for when we know there's a G or C present and one (not_GC) for when it's uncertain if there's a GC present

geno_gc <- geno8 %>%
  filter(Group %in% c("ZMMIL", "ZMMLR", "ZMMMR", "ZMPBA", "ZMPIL", "ZMPJA")) %>%
  mutate(species = fct_recode(Group, 
                              "maize" = "ZMMIL", 
                              "maize" = "ZMMLR",
                              "maize" = "ZMMMR",
                              "teosinte" = "ZMPBA",
                              "teosinte" = "ZMPIL",
                              "teosinte" = "ZMPJA"))
# filters so we take only maize and teosinte groups and adds a species column

gc_pivot <- geno_gc %>% 
  select(Group, species, GC, not_GC) %>%
  pivot_longer(., cols = c(GC, not_GC), names_to = "GC_content", values_to = "count")
# pivots GC and not_GC columns so we have a GC_content column and a count column only

gc_group_plot <- ggplot(gc_pivot, aes(fill = GC_content, y = count, x = Group)) + 
  xlab("Group") +
  ylab("Proportion") +
  geom_bar(position = "fill", stat = "identity") +
  theme(axis.text.x = element_text(angle = 90)) + # rotate text labeling groups
  scale_fill_brewer(palette = "Dark2") + # colors from RColorBrewer
  facet_grid(~species) # facet by species
gc_group_plot

ggsave("./plots/gc_group_plot.pdf", plot = gc_group_plot, width = 5, height = 5, units = "in", dpi = 300) # print to pdf

ggsave("./plots/gc_group_plot.png", plot = gc_group_plot, width = 5, height = 5, units = "in", dpi = 300) # print to png

```

I decided to make a graph showing the amount of G/G, C/C, C/G. and G/C positions (positions where it was certain that there was a G or C). To do this, I made a column counting the number of positions that had a certain G or C and then a column counting the number of positions that did not have a certain G or C. 

I filtered to include only teosinte and maize groups, then added a column for species (teosinte or maize). I then pivoted the `GC` and `not_GC` columns longer to make `GC_content` and `count` columns to count up the total positions that we were certain had a G or C and positions where there either was not a G or C or we were not certain. 

Then I made a plot showing proportion of positions where we were certain that there was a G or C compared to positions where we were not for all groups, faceted by species.

We can see that the proportion of positions where we know that there's a G or C is fairly similar for all groups, but seems slightly more variable in teosinte groups than maize groups. 

```{r}
gc_species_plot <- ggplot(gc_pivot, aes(fill = GC_content, y = count, x = species)) + 
  xlab("Species") +
  ylab("Proportion") +
  geom_bar(position = "fill", stat = "identity") +
  scale_fill_brewer(palette = "Dark2") # colors from RColorBrewer
gc_species_plot

ggsave("./plots/gc_species_plot.pdf", plot = gc_species_plot, width = 5, height = 5, units = "in", dpi = 300) # print to pdf

ggsave("./plots/gc_species_plot.png", plot = gc_species_plot, width = 5, height = 5, units = "in", dpi = 300) # print to png

```
Finally, I made a plot only showing the proportion of sites where we were certain that there was a G or C for each species, maize or teosinte. 

It appears that the proportion of cites where a G or C is certain is similar between these two species, just slightly higher in teosinte than in maize.